### PR TITLE
rustup to e7c23ab933ebc1f205c3b59f4ebc85d40f67d404

### PR DESCRIPTION
**Pull request by tmiasko**

### CODE REVIEW

File: src/shims/intrinsics.rs [hunk 1]
```diff
@@ -404,32 +404,32 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
         let this = self.eval_context_mut();
 
 
-        let &[place] = check_arg_count(args)?;
+        let &[ref place] = check_arg_count(args)?;
         let place = this.deref_operand(place)?;
 
         // make sure it fits into a scalar; otherwise it cannot be atomic
-        let val = this.read_scalar_atomic(place, atomic)?;
+        let val = this.read_scalar_atomic(&place, atomic)?;
```
Comment (RalfJung) [inline src/shims/intrinsics.rs:432]: Why is it `deref_operand(place)` a few lines above but `read_scalar_atomic(&place` here?

File: src/data_race.rs [hunk 2]
```diff
@@ -530,14 +530,14 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: MiriEvalContextExt<'mir, 'tcx> {
...
+        place: &MPlaceTy<'tcx, Tag>,
         new: ScalarMaybeUninit<Tag>,
         atomic: AtomicRwOp,
     ) -> InterpResult<'tcx, ScalarMaybeUninit<Tag>> {
         let this = self.eval_context_mut();
 
-        let old = this.allow_data_races_mut(|this| this.read_scalar(place.into()))?;
-        this.allow_data_races_mut(|this| this.write_scalar(new, place.into()))?;
+        let old = this.allow_data_races_mut(|this| this.read_scalar(&place.into()))?;
+        this.allow_data_races_mut(|this| this.write_scalar(new, &(*place).into()))?;
```
Comment (RalfJung) [inline src/data_race.rs:540]: `&place` vs `&(*place)` seems inconsistent here, they are doing the same thing, right?

Comment (tmiasko) [reply src/data_race.rs:540]: First constructs `OpTy` which has `From<&MPlaceTy>`, the other construct `PlaceTy` which doesn't have one yet (it was unnecessary for rustc).

File: src/shims/intrinsics.rs (hunk 1 above)
Comment (tmiasko) [reply src/shims/intrinsics.rs:432]: The first `ref place` binding is being shadowed by `let place = this.deref_operand(place)?;`.

**Comment by RalfJung** [comment]
Thanks a lot, this must have been a lot of work. :)

We can land this as-is, but I am confused by a few things -- see my questions above.

File: src/data_race.rs (hunk 2 above)
Comment (RalfJung) [reply src/data_race.rs:540]: Oh, I didn't realize you were adding new `From` instances as well.

File: src/shims/intrinsics.rs (hunk 1 above)
Comment (RalfJung) [reply src/shims/intrinsics.rs:432]: Oh, good point, thanks.

**Pull request merged**